wsmaster/che-core-api-workspace/src/main/java/org/eclipse/che/api/workspace/server/WorkspaceManager.java

```diff
@@ -495,6 +502,55 @@ public void createSnapshot(String workspaceId) throws NotFoundException, ServerE
         return machineManager.getSnapshots(sessionUser().getUserId(), workspaceId);
     }
 
+    /**
+     * Creates and adds new machine into specified workspace runtime.
+     * If machine creation fails, then that machine will be destroyed.
+     *
+     * @param machineConfig
+     *         configuration of new machine
+     * @param workspaceId
+     *         id of workspace this machine belongs to
+     * @param activeEnv
+     *         current machine environment
+     * @return starting machine instance
+     * Throws the same exceptions as @{@link MachineManager#createMachineAsync(MachineConfig, String, String)}
+     */
+    public MachineImpl createMachineAsyncInRuntime(MachineConfig machineConfig, String workspaceId, String activeEnv)
+            throws MachineException,
+                   BadRequestException,
+                   SnapshotException,
+                   NotFoundException,
+                   ConflictException {
+        new AddNewMachineInRuntimeEventSubscriber();
+        return machineManager.createMachineAsync(machineConfig, workspaceId, activeEnv);
+    }
+
+    /**
+     * Removes additional machine from workspace runtime.
+     *
+     * @param machineId
+     *         id of machine to remove
+     * @throws NotFoundException
+     *         when workspace from which machine is removing doesn't have runtime or
+     *         when workspace doesn't exist
+     * @throws ConflictException
+     *         when machine is dev-machine
+     * @throws MachineException
+     *         when another error occurs
+     * @throws NullPointerException
+     *         when {@code machineId} is null
+     */
+    public void removeMachineFromRuntime(String machineId) throws NotFoundException, ConflictException, MachineException {
```

Comment:
@garagatyi I disagree. I think, that we should have mechanism to remove machines from runtime.
As about wrong usage - it is responsibility of developer to use it correctly. We can broke something in many classes by invoking something wrong, isn't it?